src/main/java/net/dv8tion/jda/core/requests/Requester.java

```diff
@@ -92,65 +98,80 @@ public JDAImpl getJDA()
      *         the request can be made again. This could either be for the Per-Route ratelimit or the Global ratelimit.
      *         <br>Check if globalCooldown is {@code null} to determine if it was Per-Route or Global.
      */
-    public <T> Long execute(Request<T> apiRequest)
+    public Long execute(Request<?> apiRequest, boolean handleOnRatelimit)
     {
-        CompiledRoute route = apiRequest.getRoute();
+        Route.CompiledRoute route = apiRequest.getRoute();
         Long retryAfter = rateLimiter.getRateLimit(route);
         if (retryAfter != null)
+        {
+            if (handleOnRatelimit)
+                apiRequest.handleResponse(new Response(retryAfter, Collections.emptySet()));
             return retryAfter;
+        }
 
-        BaseRequest request;
-        Object body = apiRequest.getData();
+        okhttp3.Request.Builder builder = new okhttp3.Request.Builder();
 
-        //Special case handling for MessageChannel#sendFile.
-        // If a MultipartBody request was passed as the body then we assume it was constructed correctly
-        // and just wrap it in auth headers and allow processing.
-        if (body instanceof MultipartBody)
-        {
-            request = addHeaders((MultipartBody) body);
-        }
-        else
-        {
-            String bodyData = body != null ? body.toString() : null;
-            request = createRequest(route, bodyData);
-        }
+        String url = DISCORD_API_PREFIX + route.getCompiledRoute();
+        builder.url(url);
+
+        String method = apiRequest.getRoute().getMethod().toString();
+        RequestBody body = apiRequest.getBody();
+
+        if (body == null && HttpMethod.requiresRequestBody(method))
+            body = EMPTY_BODY;
+
+        builder.method(method, body);
+
+        builder.header("user-agent", USER_AGENT);
+
```

Comment:
This required a minor change in the code to wrap the response body in a `GZIPInputStream` when needed.